### PR TITLE
ARROW-1895/ARROW-1897: [Python] Add field_name to pandas index metadata

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by wesm** [comment]
@jorisvandenbossche would you mind reviewing and making sure this jives with the discussion so far?

**Comment by jorisvandenbossche** [comment]
One special case that I encountered in https://github.com/apache/arrow/pull/1386 is a DataFrame with column name `None` (from ipc when serializing a Series without name). 
This case is not yet handled here:

```
In [6]: pa.Table.from_pandas(pd.DataFrame({None: [1,2,3]}))
Out[6]: 
pyarrow.Table
None: int64
__index_level_0__: int64
metadata
--------
{b'pandas': b'{"index_columns": ["__index_level_0__"], "column_indexes": [{"na'
            b'me": null, "pandas_type": "mixed", "numpy_type": "object", "meta'
            b'data": null}], "columns": [{"name": null, "field_name": null, "p'
            b'andas_type": "int64", "numpy_type": "int64", "metadata": null}, '
            b'{"name": null, "field_name": "__index_level_0__", "pandas_type":'
            b' "int64", "numpy_type": "int64", "metadata": null}], "pandas_ver'
            b'sion": "0.22.0.dev0+260.g5da3759"}'}
```

So for the column, `"name": null, "field_name": null,` are both null, while field_name should be "None"

**Comment by cpcloud** [comment]
@jorisvandenbossche How is it possible in practice to get a `Series` with a name of `None`? How can one exist without explicitly constructing it or assigning its name to be `None`?

**Comment by jorisvandenbossche** [comment]
Indeed by just constructing it and not getting it as a column from a dataframe. 
Such a dataframe as I show above is constructed in the ipc code to serialize Series objects:

https://github.com/apache/arrow/blob/1d519d8351c0b26ccb3214116eaf3aa684e4bba0/python/pyarrow/serialization.py#L194-L195

It is also tested explicitly: https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_ipc.py#L458

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_convert_pandas.py]: Should we also assert that `idx0['name']` is None, or is that already tested elsewhere adequately?

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_convert_pandas.py]: For completeness, I would assert that for the other columns, the name and field_name are equal (this is obvious from the current code, but not explicitly tested):

```
col1, col2, idx0, foo = js['columns']

assert col1['name'] == col1['field_name']
assert col2['name'] == col2['field_name']
```

**Review by jorisvandenbossche — COMMENTED** [review]
Apart from the None column name comment, looks good to me! And resembles the discussion IMO.

Added some minor comments.

File: python/pyarrow/pandas_compat.py
Comment (cpcloud) [reply python/pyarrow/pandas_compat.py]: This is to preserve the `None` value when parsing `null` in JSON.

File: python/pyarrow/tests/test_convert_pandas.py
Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: I'll add that

File: python/pyarrow/tests/test_convert_pandas.py
Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: Yep, will do

File: python/pyarrow/pandas_compat.py
Comment (cpcloud) [reply python/pyarrow/pandas_compat.py]: Thanks, will do

File: python/pyarrow/pandas_compat.py
Comment (jorisvandenbossche) [reply python/pyarrow/pandas_compat.py]: Ah yes, of course. (So for columns it is converted to string when not None)

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_convert_pandas.py]: So in principle, IMO this should be the string `"None"` to make `field_name` perfectly usable for schema lookup. 
Otherwise you need to do `schema.get_field_index(field_name or "None")`

Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: Hm, this works without that, let me see what's happening.

Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: We're actually explicitly handling this case later on in `table_to_blockmanager` so this is okay.

File: python/pyarrow/pandas_compat.py
Comment (cpcloud) [reply python/pyarrow/pandas_compat.py]: done

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_convert_pandas.py]: Yes, the current code works without (just as it did work before with a "name" of None) as it is handled by `table_to_blockmanager`. 
But that means that you will always have to special case this option, and for me that should be the point of "field_name" that ``schema.get_field_index(field_name)`` is guaranteed to not error (and you thus don't have to special case None)

Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: Ok, I've implemented this. Pushing it up now.

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_convert_pandas.py]: not that important, but doing `columns= ['a', None, '__index_level_0__']` inside the `DataFrame` call is a bit simpler

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

Thanks for the updates, once this is merged, I will rebase my PR https://github.com/apache/arrow/pull/1386 against this to use the new functionality.

File: python/pyarrow/tests/test_convert_pandas.py
Comment (cpcloud) [reply python/pyarrow/tests/test_convert_pandas.py]: Yep, thank you. That is much better.

File: python/pyarrow/tests/test_convert_pandas.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_convert_pandas.py]: Tests are failing on this one. Maybe this is only the case for unicode and not for bytes, and thus only for `not PY2` ?

**Comment by wesm** [comment]
I can merge this tomorrow once the build is passing, will also take a brief look through

**Review by wesm — APPROVED** [review]
+1, let me look into the test failure

**Comment by wesm** [comment]
I'm going to be AFK for about 5 or 6 hours, @cpcloud or @xhochy please go and ahead and merge this once the Python builds run in Travis CI so @jorisvandenbossche can rebase. I'll be back on later today to work on some other patches for 0.8.0

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**
